scipy/cluster/vq.py

```diff
@@ -519,7 +529,50 @@ def _krandinit(data, k):
     return x
 
 
-_valid_init_meth = {'random': _krandinit, 'points': _kpoints}
+def _kpp(data, k):
+    """ Picks k points in data based on the kmeans++ method
+
+    Parameters
+    ----------
+    data : ndarray
+        Expect a rank 1 or 2 array. Rank 1 are assumed to describe one
+        dimensional data, rank 2 multidimensional data, in which case one
+        row is one observation.
+    k : int
+        Number of samples to generate.
+
+    Returns
+    -------
+    init : ndarray
+        A 'k' by 'N' containing the initial centroids
+
+    References
+    ----------
+    .. [1] D. Arthur and S. Vassilvitskii, "k-means++: the advantages of
+       careful seeding", Proceedings of the Eighteenth Annual ACM-SIAM Symposium
+       on Discrete Algorithms, 2007.
+    """
+
+    dims = data.shape[1] if len(data.shape) > 1 else 1
+    init = np.ndarray((k, dims))
+
+    for i in range(k):
+        if i == 0:
+            init[i, :] = data[np.random.randint(dims)]
+
+        else:
+            D2 = np.array([min(
+                            [np.inner(init[j]-x, init[j]-x) for j in range(i)]
+                            ) for x in data])
+            probs = D2/D2.sum()
+            cumprobs = probs.cumsum()
+            r = np.random.rand()
```

Comment:
This is generally not the best idea - better to instantiate a new `RandomState` object, and allow for seeding. However, the way it's done here is consistent with the other two initialization methods that are already present. And there's an explicit example using `np.random.seed`. Hence I think we should merge this as is, and open a new issue for how random numbers are generated in all these methods.

Thoughts?

Reply:
+1 for fixing the random number generator in one go

Reply:
I opened gh-9450 for this